# **AVANCE 3. BASELINE**

Nombre del Trabajo:
AI for Good: Human Heritage
Automatic pre-Hispanic to Spanish Recogniser and Translator


Nombre del asesor del proyecto: Dr. Juan Arturo Nolazco Flores

Equipo: 20
Nombre de alumnos:
 * Juan Carlos Garza Sanchez	A00821522
 * Sinaí Avalos Rivera 			A01730466

## **ONE-SHOT LEARNING**

Este código hace todo el proceso de tomar una frase en Nahuatl, usar un ejemplo de traducción para generar una traducción automática a Español utilizando el modelo Gemini-Pro de Google y luego evalúa esa traducción utilizando las métricas METEOR, BLEU y ROUGE.

In [ ]:
!pip install google-cloud-aiplatform nltk==3.8.1 rouge-score

In [ ]:
import nltk
import re

# Recursos de NLTK para tokenizacion y WordNet
nltk.download('wordnet')
nltk.download('punkt')
nltk.download('omw-1.4')

from nltk.translate.meteor_score import meteor_score
from nltk.translate.bleu_score import sentence_bleu, SmoothingFunction
from nltk.tokenize import word_tokenize
from rouge_score import rouge_scorer

import pandas as pd
import google.generativeai as genai

[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!
[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package omw-1.4 to /root/nltk_data...
[nltk_data]   Package omw-1.4 is already up-to-date!


In [ ]:
# Configurar la API de Gemini
GOOGLE_API_KEY = 'AIzaSyBMm-PD_wIbHvhFFB-6dHxYvu8tvUvf5ts'
genai.configure(api_key=GOOGLE_API_KEY)

In [ ]:
# Cargar dataset
df = pd.read_csv("hf://datasets/somosnlp-hackathon-2022/Axolotl-Spanish-Nahuatl/train.csv")
df = df[['sp', 'nah']].rename(columns={'sp': 'spanish', 'nah': 'nahuatl'})
df = df.dropna(axis=0, how='any')
df = df.astype(str)
df

/usr/local/lib/python3.11/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


spanish  \
0      Y así, cuando hizo su ofrenda de fuego, se sie...   
1      ¿Si es jade, si es oro, acaso no tendrá que ir...   
2      Y cuando el Sol estuvo solo en el cielo, enseg...   
3      Allá acudieron asimismo los señores del cabild...   
4                                                  Usos:   
...                                                  ...   
20023   El Espíritu y la esposa dicen: "¡Ven!"El que ...   
20024   Yo advierto a todo el que oye las palabras de...   
20025   y si alguno quita de las palabras del libro d...   
20026   El que da testimonio de estas cosas dice: "¡S...   
20027     La gracia de nuestro Señor Jesús sea con todos   

                                                 nahuatl  
0      Auh in ye yuhqui in on tlenamacac niman ye ic ...  
1      ¿In chalchihuitl, teocuitlatl, mach ah ca on yaz?  
2      Auh yn oyuh in yoca hualmotlalli tonatiuh ylhu...  
3      Yn oncan mohuicatza yhuan yn ciudad cabildo tl...  
4                                               Kualtia:  
...                                                  ...  
20023  On Espíritu Santo niman isihuau on Borreguito ...  
20024  Nemechtlachicahuilia nenmochimej yejhuan nenqu...  
20025  Niman tla yacaj quipopolohuilis itemachtil Dio...  
20026  On yejhuan quipantlantia yejhua in, ijquin qui...  
20027  Ma toTeco Jesucristo mechtiochihua nenmochimej...  

[20024 rows x 2 columns]

In [ ]:
# Seleccionar una frase de prueba para "One-Shot"
n = 20020
nth_row = df.iloc[[n]]
print(nth_row['nahuatl'].values[0])

Nemechtlachicahuilia nenmochimej yejhuan nenquicaquij itemachtil Dios yejhuan ipan in tlajcuilolamatl nesticaj: tla yacaj quimiyequilis in tlajtlamach, Dios quimiyequilis on plagas yejhuan ye tlajcuiloloncaj ipan in tlajcuilolamatl. 


In [ ]:
# Seleccionar un solo ejemplo de entrenamiento (one-shot)
one_shot_example = df.iloc[[100]]


Definimos la función **create_one_shot_prompt** para generar un promt que instruye al modelo a traducir la frase correspondiente en Nahuatl a Español. Se da un ejemplo de traducción en el prompt y luego se solicita al modelo que traduzca la frase objetivo.

In [ ]:
# Crear el prompt con un solo ejemplo
def create_one_shot_prompt(target_sentence, example):
    return f"""
    Translate the following Nahuatl sentence to Spanish:
    Example:
    "{example['nahuatl'].values[0]}" -> "{example['spanish'].values[0]}"

    Now, translate this sentence:
    "{target_sentence}"
    """

In [ ]:
prompt = create_one_shot_prompt(nth_row['nahuatl'].values[0], one_shot_example)

Se usa el modelo generativo gemini-pro para generar la traducción. La traducción se obtiene de la API de Gemini generando contenido a partir del prompt.

In [ ]:
# Generar la traducción con Gemini
model = genai.GenerativeModel('gemini-pro')
response = model.generate_content(prompt)

Desplegar tanto la traducción generada por el modelo como la traducción esperada (de la fila seleccionada) para compararlas.

In [ ]:
# Mostrar la traducción generada
print("Predicted:", response.text)
print("Expected:", nth_row['spanish'].values[0])

Predicted: Les exhorto, hijos míos, que guardéis los preceptos que os ha dado el verdadero Dios por medio de la revelación escrita: si exhibís regocijo, Dios os mandará las plagas y el naufragio está escrito.
Expected:  Yo advierto a todo el que oye las palabras de la profecía de este libro: Si alguno añade a estas cosas, Dios le añadirá las plagas que están escritas en este libro


Con la función normalize se toma el texto y lo convierte a minúsculas y elimina cualquier puntuación o caracteres no alfanuméricos, para ayudar a hacer las comparaciones de las métricas de manera más justa.

In [ ]:
# Función para normalizar texto
def normalize(text):
    return re.sub(r'[^\w\s]', '', text.lower())  # Quita puntuación y convierte a minúsculas

In [ ]:
# Normalizar referencia y predicción
reference = normalize(nth_row['spanish'].values[0])
hypothesis = normalize(response.text)

En este caso, estamos generando traducciones con un modelo preentrenado. Como este modelo no se entrena directamente, la preocupación de subajuste o sobreajuste se reduce, pero aún es posible evaluar la calidad de la traducción comparando los resultados obtenidos con la referencia.

Evaluación del desempeño: Dado que se está utilizando un modelo generativo, la comparación con las métricas como METEOR, BLEU, y ROUGE nos ayuda a detectar si el modelo está generando traducciones de baja calidad.

* METEOR score: Mide la calidad de las traducciones basándose en la correspondencia de n-gramas y sinónimos. El texto se tokeniza antes de calcular el score.
* BLEU score: Mide la calidad de la traducción usando n-gramas de diferentes tamaños. En este caso se utiliza el suavizado para ayudar a evitar la penalización excesiva por falta de n-gramas en las traducciones.
* ROUGE: Evalúa la calidad de la traducción en términos de precisión y recall a nivel de n-gramas y subsecuencias comúnes. 'Stemmer' ayuda a que las palabras se comparen en su forma raíz.

In [ ]:
# Calcular METEOR score
meteor = meteor_score([word_tokenize(reference)], word_tokenize(hypothesis))
print(f"METEOR score: {meteor}")

METEOR score: 0.2132847389764386


In [ ]:
# Calcular BLEU score con smoothing y BLEU-2
smoother = SmoothingFunction().method4
bleu = sentence_bleu([word_tokenize(reference)], word_tokenize(hypothesis), weights=(0.5, 0.5), smoothing_function=smoother)
print(f"BLEU score: {bleu}")


BLEU score: 0.12666009927622474


In [ ]:
# Calcular ROUGE score
scorer = rouge_scorer.RougeScorer(['rouge1', 'rouge2', 'rougeL'], use_stemmer=True)
rouge_scores = scorer.score(reference, hypothesis)
print(f"ROUGE scores: {rouge_scores}")


ROUGE scores: {'rouge1': Score(precision=0.3157894736842105, recall=0.3333333333333333, fmeasure=0.3243243243243243), 'rouge2': Score(precision=0.08108108108108109, recall=0.08571428571428572, fmeasure=0.08333333333333334), 'rougeL': Score(precision=0.21052631578947367, recall=0.2222222222222222, fmeasure=0.21621621621621623)}


En este caso, se puede observar que la traducción generada no coincide de forma esperada con la traducción esperada, lo cual es reflejado por las tres métricas de evaluación.

* **METEOR score: 0.21**
 * Este es un valor bajo, indicando que hay poca coincidencia en la correspondencia de n-gramas entre la traducción generada y la referencia. Por lo que, las traducciones podrían ser significativamente diferentes.
* **BLEU score: 0.12**
 * Tiene un valor bajo, lo que indica que la traducción generada no presenta una alta cantidad de n-gramas coincidentes con la traducción esperada.
* **ROUGE scores:**
 * **rouge1: Precision 0.3, Recall 0.33, F-measure 0.32**
 Este score mide la coincidencia de unigrams, teniendo un rendimiento moderado. Aunque se tienen algunas palabras coincidentes, la traducción generada no captura el significado de la referencia.
 * **rouge2: Precision 0.0, Recall 0.0, F-measure 0.0**
 No hay coincidencias de bigramas entre la referencia y lo obtenido, lo cual no es un buen indicador en la calidad de la traducción.
 * **rougeL: Precision 0.21, Recall 0.22, F-measure 0.21**
 Este score evalúa subsecuencias más largas, el valor obtenido muestra que hay un poco de correspondencia en las secuencias de palabras, pero no el esperado.


## **Próximas pruebas con Few-Shot Learning**

Para las siguientes entregas del proyecto, se planea experimentar con Few-Shot Leaning en vez de solo usar One-Shot. Este enfoque consiste en proporcionar al modelo múltiples ejemplos de traducción y no solo un único ejemplo como en esta entrega. Esto será con el objetivo de mejorar la capacidad del modelo para generalizar y comprender mejor el contexto.

Así que, al presentar más ejemplos de traducción, se proporciona un conjunto más amplio de datos que pueden mejorar al modelo o capturar patrones, sobre todo cuando se trabaja con vocabulario más técnico o específico sobre algún tema, como en este caso que se tienen múltiples términos religiosos.

Además, al proporcionar ejemplos adicionales, el modelo podría adaptarse mejor a las variaciones tanto de estilo de redacción como de vocabulario, haciéndolas más naturales y fieles a la referencia.

En este caso, al solo tener one-shot, un único ejemplo puede conducir al modelo a sobreajustar y generar respuestas que se alineen solo a ese ejemplo. Con múltiples ejemplos, el modelo podrá tener más opciones y así se minimizaría el sesgo hacia una única traducción.

**Desempeño mínimo esperado a obtener**

Los valores mínimos esperados de nuestras métricas para próximas iteraciones y para considerar un resultado exitoso:

* METEOR: Cercano a 0.6
* BLEU: Cercano a 0.6
* ROUGE: Se interpreta según los valores de recall y precision, y debería mejorar a medida que se afina el modelo.